### PR TITLE
[Experiment] Replace `unreachable_unchecked()` with `uninit().assume_init()` in `unwrap_unchecked()`

### PR DESCRIPTION
**Pull request by alion02**
[Relevant ACP.](https://github.com/rust-lang/libs-team/issues/378) [Relevant godbolt.](https://godbolt.org/z/WqfcT7Ys9)

Attempt to clean up the LLVM IR we generate based on the assumption that we usually don't want to `assume` the value of the discriminant (because it sticks around in the IR and randomly inhibits optimizations).

### CODE REVIEW

File: library/core/src/option.rs [hunk 1]
```diff
@@ -1037,7 +1037,7 @@ impl<T> Option<T> {
         match self {
             Some(val) => val,
             // SAFETY: the safety contract must be upheld by the caller.
-            None => unsafe { hint::unreachable_unchecked() },
+            None => unsafe { mem::MaybeUninit::uninit().assume_init() },
```
Comment (Nilstrieb) [inline library/core/src/option.rs:1040]: is this really better? that's disappointing.. how about using offset_of and accessing the value with directly with a pointer offset, entirely removing the condition?

Comment (alion02) [reply library/core/src/option.rs:1040]: I don't know if it's better in general - I just know it's better in at least one case. I think it's worth running a benchmark for it at least though. (which I can't do 🥺)

TIL about `offset_of`. `uninit` seems to be chucked pretty readily, but `offset_of` might be better in low opt level builds etc.

**Comment by workingjubilee** [comment]
Might as well try both suggested new impls of `unwrap_unchecked`.

@bors try @rust-timer queue

File: library/core/src/option.rs (hunk 1 above)
Comment (workingjubilee) [reply library/core/src/option.rs:1040]: @Nilstrieb It doesn't seem to have the desired effect:
https://godbolt.org/z/z5fv4as4s

@alion02 Please make your new libcore compile.

File: library/core/src/result.rs [hunk 2]
```diff
@@ -490,7 +490,7 @@
 
 use crate::iter::{self, FusedIterator, TrustedLen};
 use crate::ops::{self, ControlFlow, Deref, DerefMut};
-use crate::{convert, fmt, hint};
+use crate::{convert, fmt, hint, mem};
```
Comment (workingjubilee) [inline library/core/src/result.rs:493]:
```suggestion
use crate::{convert, fmt, mem};
```

File: library/core/src/option.rs (hunk 1 above)
Comment (workingjubilee) [inline library/core/src/option.rs:1040]:
```suggestion
            None => unsafe { MaybeUninit::uninit().assume_init() },
```

**Comment by the8472** [comment]
When I tried #102151 it only lead to regressions. But a lot has changed since then and this isn't the same approach so maybe the results will be different this time.

**Comment by Mark-Simulacrum** [comment]
@bors try

Generate a new try build, the old one is based on a broken master commit AFAICT.

**Comment by Mark-Simulacrum** [comment]
@rust-timer queue

**Comment by alion02** [comment]
- `nom-json` got a lot faster
- two binary size improvements for opt build
- a bunch of size regressions for debug
Much smaller impact overall than I expected, but still vaguely interesting.

By the way, I noticed that the x86 backend (which is presumably what these benchmarks are running on) is more resilient to `assume` spam than ARM. Probably something that can/should be fixed on the LLVM side, but for now — is it possible to run a benchmark on ARM?

**Comment by workingjubilee** [comment]
We don't have the tools/facilities set up for benchmarking Arm, no.

**Comment by the8472** [comment]
I assume running rustc-perf locally should work on an arm linux box.

**Comment by Nilstrieb** [comment]
any improvements from this PR are LLVM bugs, but evidently the LLVM bugs exist. you should report them upstream, but I'd be ok with a bit of papering over them in the meantime. i would prefer the offset_of based approach over this.

**Comment by JohnCSimon** [comment]
@alion02 
ping from triage - can you post your status on this PR? This PR has not received an update in a few months.


FYI: when a PR is ready for review, send a message containing 
`@rustbot ready` to switch to `S-waiting-on-review` so the PR is in the reviewer's backlog.

Or if you're not going to continue, please close it. Thank you!

**Comment by alion02** [comment]
Pursuing other priorities. May return another time. Thanks for the reminder.

**Pull request closed without merging**